// @doist/todoist-api-typescript/dist/esm/utils/validators.js
import { AttachmentSchema, SectionSchema, LabelSchema, CommentSchema, UserSchema, CurrentUserSchema, TaskSchema, PersonalProjectSchema, WorkspaceProjectSchema, ProductivityStatsSchema, ActivityEventSchema, WorkspaceUserSchema, WorkspaceInvitationSchema, WorkspacePlanDetailsSchema, JoinWorkspaceResultSchema, WorkspaceSchema, } from '../types/entities.js';
export function validateTask(input) {
    return TaskSchema.parse(input);
}
export function validateTaskArray(input) {
    return input.map(validateTask);
}
/**
 * Type guard to check if a project is a workspace project.
 * @param project The project to check
 * @returns True if the project is a workspace project
 */
export function isWorkspaceProject(project) {
    return 'workspaceId' in project;
}
/**
 * Type guard to check if a project is a personal project.
 * @param project The project to check
 * @returns True if the project is a personal project
 */
export function isPersonalProject(project) {
    return !isWorkspaceProject(project);
}
/**
 * Validates and parses a project input.
 * @param input The input to validate
 * @returns A validated project (either PersonalProject or WorkspaceProject)
 */
export function validateProject(input) {
    if ('workspaceId' in input) {
        return WorkspaceProjectSchema.parse(input);
    }
    return PersonalProjectSchema.parse(input);
}
export function validateProjectArray(input) {
    return input.map(validateProject);
}
export function validateSection(input) {
    return SectionSchema.parse(input);
}
export function validateSectionArray(input) {
    return input.map(validateSection);
}
export function validateLabel(input) {
    return LabelSchema.parse(input);
}
export function validateLabelArray(input) {
    return input.map(validateLabel);
}
export function validateComment(input) {
    return CommentSchema.parse(input);
}
export function validateCommentArray(input) {
    return input.map(validateComment);
}
export function validateUser(input) {
    return UserSchema.parse(input);
}
export function validateUserArray(input) {
    return input.map(validateUser);
}
export function validateProductivityStats(input) {
    return ProductivityStatsSchema.parse(input);
}
export function validateCurrentUser(input) {
    return CurrentUserSchema.parse(input);
}
export function validateActivityEvent(input) {
    return ActivityEventSchema.parse(input);
}
export function validateActivityEventArray(input) {
    return input.map(validateActivityEvent);
}
export function validateAttachment(input) {
    return AttachmentSchema.parse(input);
}
export function validateWorkspaceUser(input) {
    return WorkspaceUserSchema.parse(input);
}
export function validateWorkspaceUserArray(input) {
    if (!Array.isArray(input)) {
        throw new Error(`Expected array for workspace users, got ${typeof input}`);
    }
    return input.map(validateWorkspaceUser);
}
export function validateWorkspaceInvitation(input) {
    return WorkspaceInvitationSchema.parse(input);
}
export function validateWorkspaceInvitationArray(input) {
    return input.map(validateWorkspaceInvitation);
}
export function validateWorkspacePlanDetails(input) {
    return WorkspacePlanDetailsSchema.parse(input);
}
export function validateJoinWorkspaceResult(input) {
    return JoinWorkspaceResultSchema.parse(input);
}
export function validateWorkspace(input) {
    return WorkspaceSchema.parse(input);
}
export function validateWorkspaceArray(input) {
    return input.map(validateWorkspace);
}
